cpp/ql/lib/semmle/code/cpp/dataflow/ExternalFlow.qll

```diff
@@ -434,18 +434,30 @@ private predicate elementSpec(
   summaryModel(namespace, type, subtypes, name, signature, ext, _, _, _, _, _)
 }
 
+private predicate isClassConstructedFrom(Class c, Class templateClass) {
+  c.isConstructedFrom(templateClass)
+  or
+  not any(Class c_).isConstructedFrom(templateClass) and c = templateClass
+}
+
+private predicate isFunctionConstructedFrom(Function f, Function templateFunc) {
+  f.isConstructedFrom(templateFunc)
+  or
+  not any(Function f_).isConstructedFrom(templateFunc) and f = templateFunc
+}
```

Comment:
Note: I accidentially introduced a bug that meant that we match too much here. This has been fixed in 3c0af498db588c2fd3974fd64f58687b33360581

Reply:
Yeah, I found your use of `any(Class c_)` here to be suspicious, why do we care whether some third party class exists or not?  The final version is much better.

Reply:
I wouldn't mind a comment explaining what these two predicates are for, since they're named similarly to `isConstructedFrom` but do not mean the same thing.  Or possibly an expanded comment on `getFullyTemplatedFunction` - the whole thing is rather confusing regarding what can be a template, template instantiation, non-template, or combination of those possibilities.

Reply:
Fixed in 2c58279137a2f3da9b3c71e18d7966e2b9c60ad2